<h1 align="center">
<img src="https://raw.githubusercontent.com/Starlitnightly/Epiverse/main/Epiverse_guide/docs/img/logo.png" width="400">
</h1><br>

[![Lifecycle:maturing](https://img.shields.io/badge/lifecycle-maturing-blue.svg)](https://www.tidyverse.org/lifecycle/#maturing) [![bulid:passing](https://img.shields.io/appveyor/build/gruntjs/grunt)](https://img.shields.io/appveyor/build/gruntjs/grunt) [![License:GPL](https://img.shields.io/badge/license-GNU-blue)](https://img.shields.io/apm/l/vim-mode) [![pypi-badge](https://img.shields.io/pypi/v/epiverse)](https://pypi.org/project/epiverse) [![Documentation Status](https://readthedocs.org/projects/omicverse/badge/?version=latest)](https://epiverse.readthedocs.io/en/latest/?badge=latest) 

Epiverse is the fundamental package for single cell Assay for Transposase-Accessible Chromatin using sequencing (scATAC-seq) with Python. This package is in development, and is not yet ready for use.

## Directory structure

````shell
.
├── epiverse                  # Main Python package
├── epiverse_guide            # Documentation files
├── LICENSE
└── README.md
````

## Where to get it

Epiverse can be installed via pypi and you need to install `omicverse` at first. Please refer to the [installation tutorial](https://epiverse.readthedocs.io/en/latest/Installation_guild/) for more detailed installation steps and adaptations for different platforms (`Windows`, `Linux` or `Mac OS`).

You can use  `pip install -U epiverse` for installation.

## Usage

Please checkout the documentations and tutorials at [epiverse.readthedocs.io](https://epiverse.readthedocs.io/en/latest/).

## Contact

- Lei Hu ([1157269262@qq.com](mailto:1157269262@qq.com), or [hulei@westlake.edu.cn](mailto:hulei@westlake.edu.cn))
- Zehua Zeng ([starlitnightly@163.com](mailto:starlitnightly@163.com) or [zehuazeng@xs.ustb.edu.cn](mailto:zehuazeng@xs.ustb.edu.cn))
